contracts/test/invariants/Base.sol

```diff
@@ -0,0 +1,231 @@
+// 7. Write invariant_test test that does a dummy action just to see that the services are called
+// 8. run "env FOUNDRY_INVARIANT_FAIL_ON_REVERT=true FOUNDRY_INVARIANT_RUNS=128 FOUNDRY_INVARIANT_DEPTH=128 forge test -vvv --match-contract INVARIANT"
+// 9. It would be good to move Base and the logic around block.timestamp and block.number into another contract and inherit from that on the test contract. This way we can call
+//    the contract "INVARIANT_River" or something like that in order for the --match-contract argument to work
+
+contract Base is Test, BytesGenerator {
+    // Protocol contracts
+    RiverV1 public river;
+    IDepositContract public deposit;
+    WithdrawV1 public withdraw;
+    OracleV1 public oracle;
+    ELFeeRecipientV1 public elFeeRecipient;
+    CoverageFundV1 public coverageFund;
+    AllowlistV1 public allowlist;
+    OperatorsRegistryV1 public operatorsRegistry;
+    RedeemManagerV1 public redeemManager;
+
+    address internal admin;
+    address internal newAdmin;
+    address internal collector;
+    address internal newCollector;
+    address internal allower;
+    address internal oracleMember;
+    address internal newAllowlist;
+    address internal operatorOne;
+    address internal operatorOneFeeRecipient;
+    address internal operatorTwo;
+    address internal operatorTwoFeeRecipient;
+    address internal bob;
+    address internal joe;
+
+    string internal operatorOneName = "NodeMasters";
+    string internal operatorTwoName = "StakePros";
+
+    uint256 internal operatorOneIndex;
+    uint256 internal operatorTwoIndex;
+
+    uint64 constant epochsPerFrame = 225;
+    uint64 constant slotsPerEpoch = 32;
+    uint64 constant secondsPerSlot = 12;
+    uint64 constant epochsUntilFinal = 4;
+
+    uint128 constant maxDailyNetCommittableAmount = 3200 ether;
+    uint128 constant maxDailyRelativeCommittableAmount = 2000;
+
+    // Tracking data
+    uint256 timeStamp;
+    uint256 blockNumber;
+
+    // Services
+    StakerService public stakerService;
+
+    function setUp() public virtual {
+        deployProtocol();
+        deployServices();
+        addTargetSelectors();
+        excludeDeployedContracts();
+    }
+
+    function loadBlockState() public {
+        vm.warp(getTimeStamp());
+        vm.roll(getBlockNumber());
+    }
+
+    function writeBlockState() public {
+        setBlockNumber(block.number + 1);
+        setTimeStamp(block.timestamp + 1);
+    }
+
+    function getTimeStamp() public view returns (uint256) {
+        return timeStamp;
+    }
+
+    function getBlockNumber() public view returns (uint256) {
+        return blockNumber;
+    }
+
+    function setTimeStamp(uint256 _timeStamp) internal {
+        timeStamp = _timeStamp;
+    }
+
+    function setBlockNumber(uint256 _blockNumber) internal {
+        blockNumber = _blockNumber;
+    }
+
+    // @dev: This function will deploy the protocol with the correct config
+    function deployProtocol() internal {
```

Comment:
I know it's a bit cumbersome but I think it would be nice to do the depployment with the proxies to be iso with live deployments

Reply:
Makes sense, I am of the same opinion. Just wanted to get the base up & running.